source/common/upstream/cds_api_impl.cc

```diff
@@ -51,12 +47,44 @@ void CdsApiImpl::onConfigUpdate(const ResourceVector& resources, const std::stri
   for (const auto& cluster : resources) {
     MessageUtil::validate(cluster);
   }
-  // We need to keep track of which clusters we might need to remove.
+
   ClusterManager::ClusterInfoMap clusters_to_remove = cm_.clusters();
   for (auto& cluster : resources) {
-    const std::string cluster_name = cluster.name();
+    clusters_to_remove.erase(cluster.name());
+  }
+  Protobuf::RepeatedPtrField<std::string> to_remove_repeated;
+  for (auto cluster : clusters_to_remove) {
+    *to_remove_repeated.Add() = cluster.first;
+  }
+
+  Protobuf::RepeatedPtrField<envoy::api::v2::Resource> to_add_repeated;
+  for (const auto& cluster : resources) {
+    envoy::api::v2::Resource to_add;
+    to_add.set_name(cluster.name());
+    to_add.set_version(version_info);
+    ProtobufWkt::Any any;
+    any.PackFrom(cluster);
+    *to_add.mutable_resource() = any;
+    *to_add_repeated.Add() = to_add;
```

Comment:
I think it's a bit cleaner to first add, get the mutable reference, then perform the operations on that (such as `PackFrom`). That way you skip a copy.

Reply:
Done.